README.md
# {InterpretationEngine} (alpha) R package

To install the alpha version of the "interpretation engine" R package install the dependencies, then use `remotes` to get the latest version from GitHub.

```r
# install required packages 
install.packages(c("XML", "caret", "data.tree", "digest", "doParallel", "dplyr", 
                   "foreach", "plyr", "raster", "soilDB", "forcats", "tidyr", "remotes"))
                   
# install package from ncss-tech interpretation-engine repository                   
remotes::install_github('ncss-tech/interpretation-engine')
```

Input data for @josephbrehm's demonstration and other demonstrations can be found in the `inst/extdata` portion of this repository. 
A demonstration script and input boundary files can be found in `/demo` folder.

To have all that data to try, you can download the repository as a static ZIP file, or "clone" with `git` it so you can keep up with the changes that are coming!
 
# Interpretations Outside of NASIS
There are many reasons for wanting to develop, test, and apply interpretations to soil data outside the context of NASIS. This project aims to create a prototype from existing interpretation rules, evaluations, and properties as managed in NASIS. Once the prototype is complete it should be possible to generate fuzzy ratings from arbitrary sources of soil and environmental data sources.

## How Does it Work?
The [data.tree](https://cran.r-project.org/web/packages/data.tree/vignettes/data.tree.html) package defines objects and methods that are well suited to the task of describing the hierachy of rules and evaluations. The mapping between domain values and fuzzy membership can be accomplished with `approxfun`.

## Outline
1. load all rules, evaluations, properties into R via ODBC as `data.frame` objects 
3. load single rule and (recursively) load sub-rules into a `data.tree` object
4. load evaluation functions into each terminal node of `data.tree` object 
5. create wrapper function to send properties to evaluation functions
6. combine fuzzy values via operators and hedges to generate a final fuzzy rating

## Resources
1. MO-6: INTERP-Rule Tree Diagram Chart (Interactive) v1.1+

## Things to Figure Out
* convert property NASIS-report language into actual values
* link properties or insert new data
* apply logical / arith. operators to nodes
* how can we get required properties from SDA?
* some properties return an RV, some {low,RV,high}: all functions must be vectorized

## Examples

### Evaluation Curves
The following image was generated from a evaluation curve data from NASIS.

# Examples
 * http://ncss-tech.github.io/interpretation-engine/misc/nasis-interp-guts.html
 * http://ncss-tech.github.io/interpretation-engine/misc/D3/tree.html
 
![alt text](misc/static-figures/ca-storie-index-slope-eval.png)
![alt text](misc/static-figures/ca-storie-index-slope-eval-spline-0to8percent.png)


### Rule Trees
The following blocks of text are a simple representation of the `data.tree` objects that describe the hierachy of rules and evaluations that define an interpretation.

**Dust PM10 and PM2.5 Generation**
<pre style="font-size: 70%; font-family: monospace">
                                                                       levelName           Type Value RefId rule_refid eval_refid evalType
1  Dust PM10 and PM2.5 Generation                                                                                                         
2   °--RuleHedge_de473ab5                                                              multiply   0.5                                     
3       °--RuleOperator_0a397761                                                        product                                           
4           ¦--Dryness Index 0.5 to 3                                                                 18448                 18448  Sigmoid
5           °--RuleOperator_7025ab26                                                        sum                                           
6               ¦--Dust Due to Gypsum                                                                 23659      23659                    
7               ¦   °--RuleHedge_5c18cb73                                          not_null_and     0                                     
8               ¦       °--Dust from Gypsum Content 2 to 15 Percent                                   18446                 18446  Sigmoid
9               °--Dust Due to Silt and Clay                                                          23661      23661                    
10                  °--RuleHedge_caf7e2e0                                        null_not_rated     0                                     
11                      °--Dust from Silt and Clay Content 20 to 70 Percent Sand                      18447                 18447   Linear
</pre>


**CA Storie Index**

Using the `data.tree` representation of the CA Storie Index Primary Rule, we can extract the entire set of evaluations and required properties:

|   |evaluation                                                  |property                                                     |
|:--|:-----------------------------------------------------------|:------------------------------------------------------------|
|1  |*Storie Factor A Pedon Group not rated taxonomy rev         |TAXONOMIC ORDER                                              |
|2  |*Storie Factor A Pedon Group 1 and 2 soil depth             |*Storie depth to first restriction for Profile Group 1-3     |
|3  |*Storie Factor A Pedon Group 1 suborders                    |TAXONOMIC SUBORDER                                           |
|4  |*Storie Factor A Pedon Group 1 great groups                 |TAXONOMIC GREAT GROUP                                        |
|5  |*Storie Factor A Pedon Group 1 subgroups                    |TAXONOMIC SUBGROUP                                           |
|6  |*Storie Factor A Pedon Group 7, 8, 9 landform               |*Storie landforms in hills and mountains of California       |
|8  |*Storie Factor A Pedon Group 2 suborders Arents and Cambids |TAXONOMIC SUBORDER                                           |
|9  |*Storie Factor A Pedon Group 2 great groups                 |TAXONOMIC GREAT GROUP                                        |
|11 |*Storie Factor A Pedon Group 3 great groups                 |TAXONOMIC GREAT GROUP                                        |
|12 |*Storie Factor A Pedon Group 4/6 great groups               |TAXONOMIC GREAT GROUP                                        |
|13 |*Storie featkind = abrupt textural change                   |*Storie where at least one cdfkind = abrupt textural change  |
|14 |Component restriction = "abrupt textural change"            |Component restrictions                                       |
|15 |*Storie Factor A Pedon Group 3 soil depth                   |*Storie depth to first restriction for Profile Group 1-3     |
|16 |*Storie Factor A Pedon Group 5 soil depth                   |*Storie depth to pan for Profile Group 5                     |
|17 |*Storie Factor A Pedon Group 5 durids                       |TAXONOMIC SUBORDER                                           |
|18 |*Storie Factor A Pedon Group 5 great groups                 |TAXONOMIC GREAT GROUP                                        |
|19 |*Storie Factor A Pedon Group 5 subgroups                    |TAXONOMIC SUBGROUP                                           |
|21 |*Storie Factor A Pedon Group 4/6 depth abrupt text featkind |*Storie depth to featknd abrupt tex change Profile Group 4/6 |
|22 |*Storie Factor A Pedon Group 4/6 depth abrupt tex reskind   |*Storie depth to reskind abrupt tex change Profile Grp 4/6   |
|27 |*Storie Factor A Pedon Group 7 and 8 soil depth             |*Storie depth to hard bedrock/petro* for Profile Grp 7/8/9   |
|29 |*Storie Factor A Pedon Group 9 soil depth                   |*Storie depth to soft bedrock or densics for Profile Grp 7-9 |
|30 |*Storie Factor B Surface Texture                            |*Storie texture score - first horizon below duff             |
|31 |*Storie Factor B rock frag volume 0-25 cm                   |ROCK FRAG VOLUME IN DEPTH 0-25 cm OR RESTRICTION             |
|32 |*Storie Factor C Slope 0 to 100%                            |Slope, rv                                                    |
|33 |*Storie Factor X Toxicity EC maximum 0-25 cm                |EC MAXIMUM IN DEPTH 0-25 CM BELOW DUFF                       |
|34 |*Storie Factor X Toxicity SAR maximum 0-25 cm               |SAR MAXIMUM IN DEPTH 0-25 CM BELOW DUFF                      |
|35 |*Storie Factor X Toxicity pH minimum 0-25 cm                |pH MINIMUM IN DEPTH 0-25 CM BELOW DUFF                       |
|36 |*Storie Factor X Toxicity pH maximum 0-25 cm                |pH MAXIMUM IN DEPTH 0-25 CM BELOW DUFF                       |
|37 |*Storie Factor X drainage class = moderately                |GRL-DRAINAGE CLASS                                           |
|38 |*Storie component local phase is *drained*                  |**Storie component local phase is drained                    |
|39 |*Storie component local phase is *partially drained*        |**Storie component local phase is partially drained          |
|40 |*Storie Factor X drainage class = somewhat poorly           |GRL-DRAINAGE CLASS                                           |
|43 |*Storie Factor X drainage class = poor or very poorly       |GRL-DRAINAGE CLASS                                           |
|46 |*Storie Factor X drainage class = well                      |GRL-DRAINAGE CLASS                                           |
|47 |*Storie Factor X drainage class = all excessively           |GRL-DRAINAGE CLASS                                           |
|48 |*Storie Factor X landscape ponding in growing season        |*Storie ponding interaction max in growing season            |
|49 |Landscape Flooding "NONE"                                   |FLOODING FREQUENCY (Maximum Frequency)                       |
|50 |*Storie Factor X landscape flooding in growing season       |*Storie flooding interaction max in growing season           |
|52 |*Storie Factor X erosion class = 1                          |Erosion Class in component table                             |
|53 |*Storie Factor X erosion class = 2                          |Erosion Class in component table                             |
|54 |*Storie Factor X erosion class = 3                          |Erosion Class in component table                             |
|55 |*Storie Factor X erosion class = 4                          |Erosion Class in component table                             |
|56 |*Storie Factor X erosion class = 0                          |Erosion Class in component table                             |
|63 |*Storie Factor X landscape wetness, grow. season, 25-100    |*Storie near surface wetness rv depth in growing season      |
|64 |*Storie Factor X temperature regime                         |*Storie temperature regime score                             |


Here is a text-based representation of most of the `data.tree` object:
<pre style="font-size: 65%; font-family: monospace">
                                                                                              levelName           Type Value RefId rule_refid eval_refid        evalType
1   AGR - California Revised Storie Index (CA)                                                                                                                          
2    °--RuleOperator_31a1eb65                                                                                  product                                                  
3        ¦--*Storie Factor A Not Rated Soil Orders rev                                                                       49446      49446                           
4        ¦   °--RuleHedge_1193d3c1                                                                             null_or     0                                            
5        ¦       °--*Storie Factor A Pedon Group not rated taxonomy rev                                                      50481                 50481           Crisp
6        ¦--*Storie Factor A                                                                                                 15976      15976                           
7        ¦   °--RuleOperator_01948bb8                                                                               or                                                  
8        ¦       ¦--*Storie Factor A Profile Group 1 rev                                                                     49448      49448                           
9        ¦       ¦   °--RuleOperator_a72ba798                                                                      and                                                  
10       ¦       ¦       ¦--RuleOperator_b267bf8e                                                                times                                                  
11       ¦       ¦       ¦   ¦--*Storie Factor A Profile Group 1 and 2 fuzzy depth                                           15967      15967                           
12       ¦       ¦       ¦   ¦   °--RuleHedge_7fad1a79                                                    not_null_and     0                                            
13       ¦       ¦       ¦   ¦       °--*Storie Factor A Pedon Group 1 and 2 soil depth                                      12916                 12916  ArbitraryCurve
14       ¦       ¦       ¦   °--*Storie Factor A Profile Group 1 taxonomy rev                                                49447      49447                           
15       ¦       ¦       ¦       °--RuleOperator_4d1fdd8f                                                           or                                                  
16       ¦       ¦       ¦           ¦--RuleHedge_44b2d8f3                                                not_null_and     0                                            
17       ¦       ¦       ¦           ¦   °--*Storie Factor A Pedon Group 1 suborders                                         12914                 12914           Crisp
18       ¦       ¦       ¦           ¦--RuleHedge_f2a5b4ba                                                not_null_and     0                                            
19       ¦       ¦       ¦           ¦   °--*Storie Factor A Pedon Group 1 great groups                                      12915                 12915           Crisp
20       ¦       ¦       ¦           °--RuleHedge_25e606a5                                                not_null_and     0                                            
21       ¦       ¦       ¦               °--*Storie Factor A Pedon Group 1 subgroups                                         13092                 13092           Crisp
22       ¦       ¦       °--RuleHedge_ea4b1839                                                                     not     0                                            
23       ¦       ¦           °--*Storie Factor A Profile Group 7,8,9 landform                                                15974      15974                           
24       ¦       ¦               °--RuleHedge_608f03c8                                                    not_null_and     0                                            
25       ¦       ¦                   °--*Storie Factor A Pedon Group 7, 8, 9 landform                                        12917                 12917           Crisp
26       ¦       ¦--*Storie Factor A Profile Group 2 rev                                                                     49428      49428                           
27       ¦       ¦   °--RuleOperator_8a20de79                                                                      and                                                  
28       ¦       ¦       ¦--RuleOperator_71911310                                                                times                                                  
29       ¦       ¦       ¦   ¦--*Storie Factor A Profile Group 1 and 2 fuzzy depth                                           15967      15967                           
30       ¦       ¦       ¦   ¦   °--RuleHedge_7fad1a79                                                    not_null_and     0                                            
31       ¦       ¦       ¦   ¦       °--*Storie Factor A Pedon Group 1 and 2 soil depth                                      12916                 12916  ArbitraryCurve
32       ¦       ¦       ¦   °--*Storie Factor A Profile Group 2 taxonomy                                                    15991      15991                           
33       ¦       ¦       ¦       °--RuleOperator_0ad34d47                                                           or                                                  
34       ¦       ¦       ¦           ¦--RuleHedge_698c3a67                                                not_null_and     0                                            
35       ¦       ¦       ¦           ¦   °--*Storie Factor A Pedon Group 2 suborders Arents and Cambids                      12929                 12929           Crisp
36       ¦       ¦       ¦           °--RuleHedge_8fbcb397                                                not_null_and     0                                            
37       ¦       ¦       ¦               °--*Storie Factor A Pedon Group 2 great groups                                      12927                 12927           Crisp
38       ¦       ¦       °--RuleHedge_ea4b1839                                                                     not     0                                            
39       ¦       ¦           °--*Storie Factor A Profile Group 7,8,9 landform                                                15974      15974                           
40       ¦       ¦               °--RuleHedge_608f03c8                                                    not_null_and     0                                            
41       ¦       ¦                   °--*Storie Factor A Pedon Group 7, 8, 9 landform                                        12917                 12917           Crisp
42       ¦       ¦--*Storie Factor A Profile Group 3                                                                         16017      16017                           
43       ¦       ¦   °--RuleOperator_a84c18cd                                                                      and                                                  
44       ¦       ¦       ¦--*Storie Factor A Profile Group 3 taxonomy                                                        15992      15992                           
45       ¦       ¦       ¦   °--RuleOperator_c859411b                                                               or                                                  
46       ¦       ¦       ¦       ¦--RuleHedge_a14e5a0e                                                    not_null_and     0                                            
47       ¦       ¦       ¦       ¦   °--*Storie Factor A Pedon Group 3 great groups                                          12931                 12931           Crisp
48       ¦       ¦       ¦       °--RuleOperator_15a86db4                                                          and                                                  
49       ¦       ¦       ¦           ¦--RuleHedge_da245a09                                                not_null_and     0                                            
50       ¦       ¦       ¦           ¦   °--*Storie Factor A Pedon Group 4/6 great groups                                    12934                 12934           Crisp
51       ¦       ¦       ¦           ¦--RuleHedge_61008193                                                         not     0                                            
52       ¦       ¦       ¦           ¦   °--RuleHedge_1a2fddac                                            not_null_and     0                                            
53       ¦       ¦       ¦           ¦       °--*Storie featkind = abrupt textural change                                    13106                 13106           Crisp
54       ¦       ¦       ¦           °--RuleHedge_214d73a9                                                         not     0                                            
55       ¦       ¦       ¦               °--RuleHedge_137ed382                                            not_null_and     0                                            
56       ¦       ¦       ¦                   °--Component restriction = "abrupt textural change"                             11459                 11459           Crisp
57       ¦       ¦       °--*Storie Factor A Profile Group 3 fuzzy depth                                                     16016      16016                           
58       ¦       ¦           °--RuleHedge_725c57d0                                                        not_null_and     0                                            
59       ¦       ¦               °--*Storie Factor A Pedon Group 3 soil depth                                                12938                 12938  ArbitraryCurve
60       ¦       ¦--*Storie Factor A Profile Group 5                                                                         16038      16038                           
61       ¦       ¦   °--RuleOperator_38a12238                                                                      and                                                  
62       ¦       ¦       ¦--RuleOperator_1ffc51b2                                                                times                                                  
63       ¦       ¦       ¦   ¦--*Storie Factor A Profile Group 5 fuzzy depth                                                 16035      16035                           
64       ¦       ¦       ¦   ¦   °--RuleHedge_a4df1747                                                    not_null_and     0                                            
65       ¦       ¦       ¦   ¦       °--*Storie Factor A Pedon Group 5 soil depth                                            12939                 12939  ArbitraryCurve
66       ¦       ¦       ¦   °--*Storie Factor A Profile Group 5 taxonomy                                                    16037      16037                           
67       ¦       ¦       ¦       °--RuleOperator_d8ea2331                                                           or                                                  
68       ¦       ¦       ¦           ¦--RuleHedge_aba67333                                                not_null_and     0                                            
69       ¦       ¦       ¦           ¦   °--*Storie Factor A Pedon Group 5 durids                                            12932                 12932           Crisp
70       ¦       ¦       ¦           ¦--RuleHedge_a831076b                                                not_null_and     0                                            
71       ¦       ¦       ¦           ¦   °--*Storie Factor A Pedon Group 5 great groups                                      12933                 12933           Crisp
72       ¦       ¦       ¦           °--RuleHedge_7954eb2b                                                not_null_and     0                                            
73       ¦       ¦       ¦               °--*Storie Factor A Pedon Group 5 subgroups                                         13093                 13093           Crisp
74       ¦       ¦       °--RuleHedge_ea4b1839                                                                     not     0                                            
75       ¦       ¦           °--*Storie Factor A Profile Group 7,8,9 landform                                                15974      15974                           
76       ¦       ¦               °--RuleHedge_608f03c8                                                    not_null_and     0                                            
77       ¦       ¦                   °--*Storie Factor A Pedon Group 7, 8, 9 landform                                        12917                 12917           Crisp
78       ¦       ¦--*Storie Factor A Profile Group 4/6                                                                       16040      16040                           
79       ¦       ¦   °--RuleOperator_867650f1                                                                    times                                                  
80       ¦       ¦       ¦--*Storie Factor A Profile Group 4/6 fuzzy depth                                                   16041      16041                           
81       ¦       ¦       ¦   °--RuleOperator_804837c6                                                              and                                                  
82       ¦       ¦       ¦       ¦--RuleHedge_74ec3ef6                                                    not_null_and     0                                            
83       ¦       ¦       ¦       ¦   °--*Storie Factor A Pedon Group 4/6 depth abrupt text featkind                          13207                 13207  ArbitraryCurve
84       ¦       ¦       ¦       °--RuleHedge_5d562f5b                                                    not_null_and     0                                            
85       ¦       ¦       ¦           °--*Storie Factor A Pedon Group 4/6 depth abrupt tex reskind                            13208                 13208  ArbitraryCurve
86       ¦       ¦       °--*Storie Factor A Profile Group 4/6 taxonomy w/ abrupt text                                       15996      15996                           
87       ¦       ¦           °--RuleOperator_d7eb1bf2                                                              and                                                  
88       ¦       ¦               ¦--*Storie Factor A Pedon Group 4/6 great groups                                            12934                 12934           Crisp
89       ¦       ¦               °--RuleOperator_55f9f74b                                                           or                                                  
90       ¦       ¦                   ¦--*Storie featkind = abrupt textural change                                            13106                 13106           Crisp
91       ¦       ¦                   °--Component restriction = "abrupt textural change"                                     11459                 11459           Crisp
92       ¦       °--*Storie Factor A Profile Groups 7, 8 or 9                                                                15975      15975                           
93       ¦           °--RuleOperator_e9966d4d                                                                      and                                                  
94       ¦               ¦--RuleOperator_48460ff8                                                                times                                                  
95       ¦               ¦   ¦--*Storie Factor A Profile Group 7,8,9 landform                                                15974      15974                           

NEWS.md
# InterpretationEngine 0.0.0.9001

* Arbitrary curves now use the internal NASIS spline algorithm to generate control points and `splinefun()` to interpolate ratings 
* `plotEvaluation()` takes an optional `pch` argument to control the plotting character for control points; control points are stored in attributes of the interpolator function: attributes `"domain"` and `"range"` are X and Y values, respectively
* Fixed sigmoid rating curve scaling
* Added support for crisp expression functions
* Added `lookupRatingClass()` method 
* Added various hedge and operator functions for testing
* Added a `NEWS.md` file to track changes to the package.


misc/new.R
library(InterpretationEngine)

# load cached data
load('cached-NASIS-data.Rda')

y <- rules[rules$rulename == 'FOR - Road Suitability (Natural Surface)', ]

dt <- parseRule(y)

# recusively splice-in sub-rules
dt$Do(traversal = 'pre-order', fun = linkSubRules)

# splice-in evaluation functions, if possible
dt$Do(traversal = 'pre-order', fun = linkEvaluationFunctions)

# # print more attributes
# options(width=300)
# print(dt, 'Type', 'Value', 'RefId', 'rule_refid', 'eval_refid', 'evalType', limit=25)
# 
# print(dt, 'Type', 'Value', 'evalType', 'propname', 'propiid', 'propuom', limit=25)

# bad <- subset(evals, is.na(propmin) & is.na(propmax))
# bad2 <- do.call('rbind', lapply(split(bad, 1:nrow(bad)), function(x) {
#             InterpretationEngine:::extractCrispExpression(x$eval, asString = TRUE)
#           }))
# View(data.frame(a=bad2))

e <- evals[evals$evaliid == 12765,]
f.INTERPRETABLE_COMPONENT <- InterpretationEngine:::extractCrispExpression(e$eval, invert = e$invertevaluationresults)
coiids <- c("642626", "1498364", "1154650", "627824", "1693178", "1419756", 
            "666236", "666657", "720510", "1270409", "642168", "1693202", 
            "635415", "659217", "1451814", "1976422", "2714408", "666222", 
            "1434478", "642641", "642166", "2029004", "720642", "1498380", 
            "642172", "2669525", "666363", "1562817", "645899", "628609", 
            "641731", "641757", "696370", "1498357", "666570", "628644", 
            "642829", "666227", "938589", "666064", "1169541", "2292963", 
            "712188", "666183", "2775950", "642445", "643048", "665515", 
            "666188", "641728")
component_property <- lookupProperties(coiids, e$propiid)

dt$RuleOperator_369952a0$Get("Type")[1]

print(dt$RuleOperator_369952a0$`Component Kind Misc. and CHIID is null`, 
      'Type', 'Value', 'evalType', 'propname', 'propiid', 'propuom', limit=25)

# this property only returns a compkind value for interpretable component kinds
#  for interpretable components the expected value is "0" and non-interpretable "Not rated"
InterpretationEngine:::.NULL_NOT_RATED(f.INTERPRETABLE_COMPONENT(component_property$rv))
InterpretationEngine:::.NULL_NOT_RATED(dt$RuleOperator_369952a0$`Component Kind Misc. and CHIID is null`$children$RuleHedge_7314c2a3$children$`Component Kind Misc. (Null)`$Get('evalFunction', component_property$rv))

node <- dt$RuleOperator_369952a0$children[[1]]
cp <- lookupProperties(coiids, na.omit(node$Get('propiid')))
cp$rv

rc <- InterpretationEngine:::lookupRatingClass(max(na.omit(node$Get('rule_refid'))))

node$children$`Slope <6% to >12%`$children$RuleHedge_c9d4901f$children$`Slope <6% to >12%`$Get('evalFunction', cp$rv)[,1] |> 
  factor(levels = rc$ratingclassupperboundary, labels = rc$ratingclassname)

node <- dt$RuleOperator_369952a0$children[[2]]
node |> print("Type")
pids <- na.omit(node$Get('propiid'))
cp <- lapply(pids, function(x) lookupProperties(coiids, x))
names(cp) <- pids
rids <- na.omit(node$Get('rule_refid'))
rc <- lapply(rids, function(x) InterpretationEngine:::lookupRatingClass(x))

node$children$RuleOperator_131c5cf9$`Fragments <15% to >50% Surface Cover (>=75 & <250mm)`$children$RuleHedge_991b51b6$children[[1]]$Get('evalFunction', cp[[3]]$rv)[,1] |> 
  factor(levels = rc[[3]]$ratingclassupperboundary, labels = rc[[3]]$ratingclassname)

node <- dt$RuleOperator_369952a0$children[[3]]
node |> print("Type")
cp <- lookupProperties(coiids, na.omit(node$Get('propiid')))
node$Get('evalFunction', cp$rv)$`Plasticity Index >=30 Thickest Layer in Depth 0-15cm, Rev`

node <- dt$RuleOperator_369952a0$children[[4]]
node |> print("Type")
cp <- lookupProperties(coiids, na.omit(node$Get('propiid')))
node$Get('evalFunction', cp$rv)$`Particles 0 to >85% Coarser vfs, >=7cm Thck, 0-15cm Dp, Rev`

node <- dt$RuleOperator_369952a0$children[[5]]
node |> print("Type")
cp <- lookupProperties(coiids, na.omit(node$Get('propiid')))
node1 <- node$RuleOperator_43c9b640$children[[1]]
cp <- lookupProperties(coiids, na.omit(node1$Get('propiid')))
node1$`Unified (Inorganic) In Layer >=7cm Thick In Depth 0-15cm`$RuleHedge_6291e31f$`Unified (Inorganic) In Layer >=7cm Thick In Depth 0-15cm`$Get('evalFunction', cp$rv)
node2 <- node$RuleOperator_43c9b640$children[[2]]
cp <- lookupProperties(coiids, na.omit(node2$Get('propiid')))
node2$`Unified (Organic) In Layer >=7cm Thick In Depth 0-15cm`$Get('evalFunction', cp$rv)

node <- dt$RuleOperator_369952a0$children[[6]]
node |> print("Type")
node1 <- node$RuleOperator_951eaacf$children[[1]]
cp <- lookupProperties(coiids, na.omit(node1$Get('propiid')))
node1$Get('evalFunction', cp$rv)$`Months With Frequent Ponding`
node2 <- node$RuleOperator_951eaacf$children[[2]]
cp <- lookupProperties(coiids, na.omit(node2$Get('propiid')))
node2$Get('evalFunction', cp$rv)$`Months With Occasional Ponding`

node <- dt$RuleOperator_369952a0$children[[7]]
node |> print("Type")
node1 <- node$RuleOperator_088b4dbf$children[[1]]
cp <- lookupProperties(coiids, na.omit(node1$Get('propiid')))
node1$Get('evalFunction', cp$rv)$`Months With Frequent or Very Frequent Flooding`
node2 <- node$RuleOperator_088b4dbf$children[[2]]
cp <- lookupProperties(coiids, na.omit(node2$Get('propiid')))
node2$Get('evalFunction', cp$rv)$`Months With Occasional Flooding`

node <- dt$RuleOperator_369952a0$children[[8]]
node |> print("Type")
node1 <- node$RuleOperator_cde17137$children[[1]]
cp <- lookupProperties(coiids, na.omit(node1$Get('propiid')))
node1$Get('evalFunction', cp$rv)$`Soil Slippage Potential, Revised`
node2 <- node$RuleOperator_cde17137$children[[2]]
cp <- lookupProperties(coiids, na.omit(node2$Get('propiid')))
node2$Get('evalFunction', cp$rv)$`Soil Slippage Potential Slope Percent 5 to 25`

node <- dt$RuleOperator_369952a0$children[[9]]
node |> print("Type")
cp <- lookupProperties(coiids, na.omit(node$Get('propiid')))
node$Get('evalFunction', cp$rv)$`Depth to Water Table Average for Consecutive Months`

node <- dt$RuleOperator_369952a0$children[[10]]
node |> print("Type")
cp <- lookupProperties(coiids, na.omit(node$Get('propiid')))
node$Get('evalFunction', cp$rv)$`Component Kind Misc. (Null)`

node <- dt$RuleOperator_369952a0$children[[11]]
node |> print("Type")
cp <- lookupProperties(coiids, na.omit(node$Get('propiid')))
node$Get('evalFunction', cp$rv)$`Null Horizon Data`

node <- dt$RuleOperator_369952a0$children[[12]]
node |> print("Type")
node1 <- node$RuleHedge_de473ab5$RuleOperator_0a397761$children[[1]]
node2 <- node$RuleHedge_de473ab5$RuleOperator_0a397761$children[[2]]
node3 <- node2$children[[1]]
node4 <- node2$children[[2]]

# oh these property results are rounded off to integer values...
cp1 <- lookupProperties(coiids, na.omit(node1$Get('propiid')))

node1$Get('evalFunction', cp1$rv)
plotEvaluation(subset(evals, evalname == "Dryness Index 0.5 to 3"), xlim=c(0,3))
v <- extractEvalCurve(subset(evals, evalname == "Dryness Index 0.5 to 3"))
v(c(0.5, 0.75, 2.75, 3))
v(c(1.098901099)) # close to expected value of 0.115
#> [1] 0.1158786

cp3 <- lookupProperties(coiids, na.omit(node3$Get('propiid')))
node3$Get('evalFunction', cp3$rv)$`Dust from Gypsum Content 2 to 15 Percent`
plotEvaluation(subset(evals, evalname == "Dust from Gypsum Content 2 to 15 Percent"), xlim=c(0,20))

cp4 <- lookupProperties(coiids, na.omit(node4$Get('propiid')))
node4$Get('evalFunction', cp4$rv)$`Dust from Silt and Clay Content 20 to 70 Percent Sand`
plotEvaluation(subset(evals, evalname == "Dust from Silt and Clay Content 20 to 70 Percent Sand"), xlim=c(20,70))

node3$Get('evalFunction', cp3$rv)$`Dust from Gypsum Content 2 to 15 Percent` + node4$Get('evalFunction', cp4$rv)$`Dust from Silt and Clay Content 20 to 70 Percent Sand` * node1$Get('evalFunction', cp1$rv) * 0.5

node <- dt$RuleOperator_369952a0$children[[13]]
node |> print("Type")
cp <- lookupProperties(coiids, na.omit(node$Get('propiid')))
node$Get('evalFunction', cp$rv)$`Drainage Class is Not Subaqueous`

